docs/_posts/2022-05-06-writing-cli-tools-with-typescript.md

```diff
@@ -0,0 +1,194 @@
+
+```bash
+git clone git@github.com:fkurz/typescript-cliutil-starter.git my-cli
+cd my-cli 
+npm i
+```
+
+Our starter project typescript-cliutil-starter uses 
+
+* the [*commander*](https://www.npmjs.com/package/commander) package for argument parsing and executing commands, 
+
+* code linting and formatting with [typescript-eslint](https://github.com/typescript-eslint/typescript-eslint) and [prettier](https://prettier.io/), 
+* [Jest](https://jestjs.io/) as test runner as well as testing and assertion library, and 
+* [pkg](https://github.com/vercel/pkg) to build a stand alone binary.
+
+
+
+To illustrate the development flow, we now add an example subcommand and test it. 
+
+For example, we could add a command _yell_ which is exactly like the *say* command (that already exists in the project) but a bit less subtle.
+
+To do this, we add a file _src/yell.ts_ and then add the following code.
+
+```typescript
+// src/yell.ts
+import { Command } from "commander";
+
+export default (): Command =>
+  new Command()
+    .name("yell")
+    .description("Say the word passed as the first argument but louder")
+    .argument("<word>")
+    .action((word: string) => console.log(word.toLocaleUpperCase()));
+```
+
+We also have to register our yell command with the main CLI command in src/cmd.ts, i.e. add an import statement for our subcommand and register it using *Commander.addCommand* 
+
+```typescript
+// src/cmd.ts
+// … 
+import buildYellCmd from "./yell";
+// …
+export default (): Command => {
+  const command = new Command()
+    .option("-g, --greet", `Say ${HELLO}`, false)
+    .addCommand(buildSayCmd())
+    .addCommand(buildYellCmd())
+    .addHelpCommand()
+    .showHelpAfterError();
+```
+
+Having done that, we can run a quick test on our command using the `dev` script in package.json. The dev script simply executes our TypeScript code with    [`ts-node`](https://github.com/TypeStrong/ts-node).
+
+```
+$ npm run dev -- yell 'hey!'
+
+> @fkurz/typescript-cliutil-starter@1.0.0 dev /private/tmp/my-cli
+> ts-node src/main.ts "yell" "hey!"
+
+HEY!
+```
+
+We can also add a quick test by adding _src/yell.test.ts_ wit the following content
+
+```typescript
+import buildCmd from "./cmd";
+
+beforeEach(() => {
+  jest.spyOn(process, "exit").mockImplementation();
+  jest.spyOn(console, "log").mockImplementation();
+});
+
+const buildArgs = (...args: string[]) => ["node", "cmd", ...args];
+
+describe("Yell subcommand", () => {
+  it("Should say 'HEY!' when passed 'hey!'", () => {
+    const cmd = buildCmd();
+    cmd.parse(buildArgs("yell", "hey!"));
+
+    expect(console.log).toHaveBeenCalledWith("HEY!");
+  });
+});
+```
+
+Since our Jest configuration expects JavaScript test files, we first have to transpile our code with the build script using the TypeScript compiler *tsc*. The transpiled JavaScript source code is put into the dist/ folder.
```

Comment:
Normally you would use something like [ts-jest](https://www.npmjs.com/package/ts-jest) to avoid this part.

As an alternative, you could also use [vitest](https://vitest.dev/), which is very similar to `jest`, but supports typescript out of the box, without any configuration.

Reply:
Ok sure, makes sens. I wanted to keep things simple here though. ts-jest (or vitest) probably would require additional configuration and likely doesn't do anything else than transpiling and running test on the transpiled code anyway.